### DISCUSSION TITLE
Do we expect projects to pin their GHC20XX version / are breaking changes in GHC20XX upgrades okay?

### DISCUSSION THREAD

**Discussion by maralorn** [ghc-proposals/ghc-proposals#432] [General]
(I hope it’s okay, that I am opening a discussion here. I didn‘t want to derail the mentioned PR with this off-topic discussion.)

This is a reply to @vdukhovni in https://github.com/ghc-proposals/ghc-proposals/pull/302#issuecomment-889490506:

> My concern about overloading the existing `-XLambdaCase` is that previously this could have been a candidate for enabling it by default since `case` is a reserved word. But now that becomes much more difficult. The proposed alternative `-XLambdaCases` that subsumes `-XLambdaCase` would not preclude making `-XLambdaCase` enabled by default in `-XGHC20XX`.

I am uncertain whether this is correct conclusion. I think we should expect every project to explicitly declare a fixed GHC20XX version it uses. (Much like today every .cabal contains either Haskell98 or Haskell2010.)
That means upgrading from GHC2021 to GHC202X is a manual, supervised upgrade that in my opinion should be allowed to break code with infrequent corner cases. 

This is not an arbitrary opinion, but the way Rust editions work. I am not sure that there is a consensus that GHC20XX should work this way, too. But I really hope we will do it that way, because it will give people the stability guarantees they need. Their projects will continue to work with future ghc versions. While giving us the ability to innovate the "default Haskell".

**Comment by vdukhovni** [comment]
The latest [GHC User's Guide](https://downloads.haskell.org/ghc/9.2.1-alpha2/docs/html/users_guide/exts/control.html) says:
```
    GHC2021

        GHC blesses a number of extensions, beyond Haskell 2010, to be
        suitable to turned on by default. These extensions are
        considered to be stable and conservative.

        GHC2021 is used by GHC if neither Haskell98 nor Haskell2010 is
        turned on explicitly. Since later versions of GHC may use a
        later GHC20xx by default, users are advised to declare the
        language set explicitly with -XGHC2021.
```
So, on the one hand, indeed users are encouraged to be explicit about the desired language version, but on the other hand the default version changes as time goes by.

I don't know what fraction of projects leave the language version implicit.

Also, Cabal 3.4 uses "Haskell2010" by default, so the implicit behaviour also eventually changes for Cabal projects: https://github.com/haskell/cabal/issues/6288

**Reply by phadej** [reply]
Omitting `default-languages` makes compiler uses its default. I.e. it doesn't default `default-language` to `Haskell2010`.

**Reply by vdukhovni** [reply]
> Omitting `default-languages` makes compiler uses its default. I.e. it doesn't default `default-language` to `Haskell2010`.

I think this proposal is arguing for something that facilitates reproducible builds, and therefore cabal behaviour that is consistent over time, but for that one probably needs stack with a fixed "resolver" version.

In cabal, one would have to freeze (perhaps module patch level) the compiler version and most of the dependencies explicitly. At which point the default language is also effectively frozen...

**Comment by nomeata** [comment]
My intention here is certainly that people pin their `GHC20XX` in projects.

The point of a changing default is that we want to give people opening a blank `ghci` session, or running experiments in plain `ghc` a modern experience at a reasonable convenience. But for “projects” with Cabal files etc. clearly such changes in behavior are not desirable.

**Comment by phadej** [comment]
What you mean by "project". As far as `cabal-install` or `stack` projects go, you cannot define global "default language" as it's (99% likely) specified by each of the included `.cabal` files in their "project" (including dependencies). However, the default language *is* defined on that level (and may be different in different packages based on what maintainers of the dependencies see fit).

Without telling what is the "project", this discussion will go sideways (as it feels it did in @vdukhovni reply to my previous comment).

Maybe there is confusion some that GHC2020 is somehow global setting, for example affecting ABI. That is **not** the case.

**Reply by nomeata** [reply]
With “project” I confusingly meant “something worth writing declarative information for”. I.e. stuff that has at least a `.cabal` file or something similar.

**Reply by phadej** [reply]
I was referring to 

>  I think we should expect every project to explicitly declare a fixed GHC20XX

In @maralorn opening post.

**Reply by maralorn** [reply]
Ah, well. That is a good question. I think I basically meant everything that has it's own cabal file. But I don't think it's that important. What I truly meant is "any unit of code that is supposed to work in the future."

I think we should say something like "If you want your code to reliably compile with future ghc versions you should pin your GHC20XX flavor." (I am not sure that we can actually give that stability guarantee but we should do our best.)

How people pin the version does not really matter. I would certainly suggest using that cabal file. But you can do it also per file and it off course does not have to be consistent for the whole project.

I want GHC to be able to progress. Extensions should be allowed to steal a bit of syntax now and then and should be possible to get removed again from GHC20XX when better solutions have emerged.
But on the same time from a packager perspective untouched code should break as rarely as possible. That's why I strongly hope that e.g. all packages on hackage will pin the Haskell flavor(s) they use. 

I know that cabal-install recently made the default-language field optional. I hope the field will stay in the file generated by cabal init and in examples and templates throughout the ecosystem.
